### ISSUE TITLE
Bug: False positive `no-useless-escape` in regular expression literal, when it includes `\h`

### ISSUE THREAD

**Issue by racz16** [eslint/eslint#17785]
### Environment

Node version: 18.6.0
npm version: 8.13.2
Local ESLint version: 8.40.0
Global ESLint version: -
Operating System: Windows 11 22H2

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

```
/**@type {import('eslint').Linter.Config} */
// eslint-disable-next-line no-undef
module.exports = {
	root: true,
	parser: '@typescript-eslint/parser',
	plugins: ['@typescript-eslint'],
	extends: ['eslint:recommended', 'plugin:@typescript-eslint/recommended'],
	rules: {
		semi: [2, 'always'],
		'@typescript-eslint/no-unused-vars': 0,
		'@typescript-eslint/no-explicit-any': 0,
		'@typescript-eslint/explicit-module-boundary-types': 0,
		'@typescript-eslint/no-non-null-assertion': 0,
	},
};

```
</details>

```ts
const regExp = /\h/;
```

### What did you expect to happen?

ESLint doesn't find any problems.

### What actually happened?

ESLint thinks that `\h` is an unnecessary character escape, however, in a regular expression, it's a horizontal whitespace shorthand character class. Output:
> error  Unnecessary escape character: \h  no-useless-escape

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiY29uc3QgcmVnRXhwID0gL1xcaC87Iiwib3B0aW9ucyI6eyJlbnYiOnsiZXM2Ijp0cnVlfSwicnVsZXMiOnsibm8tdXNlbGVzcy1lc2NhcGUiOlsiZXJyb3IiXX0sInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fSwiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0In19fQ==

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by mdjermanovic** [comment]
Hi @racz16, thanks for the issue! 

> ESLint thinks that `\h` is an unnecessary character escape, however, in a regular expression, it's a horizontal whitespace shorthand character class

In JavaScript regular expressions, `\h` is IdentityEscape (or syntax error in UnicodeMode) so the rule works as intended.

```js
/\h/.test(" "); // false
/\h/.test("h"); // true
/\h/u; // syntax error
/\h/v; // syntax error
```
